### DISCUSSION TITLE
BuildLegacyAPK unrecognized element with Java 7 ADT

### DISCUSSION THREAD

**Discussion by SUSDEV** [airsdk/Adobe-Runtime-Support#1721] [Q&A] [answered]
I have an old app that I've been updating through the years and need to build with Java 7 because the certificate is 1024 bit not 2048 bit. Harman provided the mechanism in the past to use Java 7 in a similar form to adt.bat: `@"C:\sdks\jre7\bin\java" -jar "%~dp0\..\lib\adt.jar" %*`
With 33.1.1.743, I added an element to the XML manifest `<BuildLegacyAPK>true</BuildLegacyAPK>` according to the Release Notes. However, when I try using the adt for Java 7, I get an error

`error 103: application.android.BuildLegacyAPK is an unexpected element/attribute`

anyone know to resolve this? My next course of action is to try to update my certificate to 2048 bit [according to here](https://github.com/airsdk/Adobe-Runtime-Support/discussions/1611), but going to aab seems like a lot of work.

**Comment by ajwfrost** [comment, marked as answer]
Hi - the `application.android.BuildLegacyAPK` element will only be recognised if your application descriptor file is set to a namespace of 33.1. Hopefully it would be possible to just change that, at the start of the XML file, and the rest of the process can stay the same...

If you do run into issues with it not liking Java 7, please let us know: there's also a version of two of Java 8 that still support 1024-bit certificates, and we'd seen some of the Android tools needing to use Java 8 so those ones (e.g. 1.8.0_41) may be needed for this.

The new App Bundle mechanism can also be used I believe, because app bundles don't actually need to be signed directly, but instead there are upload certificates that can be used separately from the app signing certificate... see https://airsdk.dev/docs/tutorials/platform/android/packaging-android-app-bundles#signing-and-certificates 

thanks

**Reply by SUSDEV** [reply]
Thanks for your quick reply. I tried changing the namespace and it got farther, but returned this error. Any thoughts about whether it can be remedied with what I have without going to the next step of Java 8?

```
Exception in thread "main" java.lang.UnsupportedClassVersionError: com/android/apksig/ApkSigner$SignerConfig$Builder : Unsupported major.minor version 52.0
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(Unknown Source)
        at java.security.SecureClassLoader.defineClass(Unknown Source)
        at java.net.URLClassLoader.defineClass(Unknown Source)
        at java.net.URLClassLoader.access$100(Unknown Source)
        at java.net.URLClassLoader$1.run(Unknown Source)
        at java.net.URLClassLoader$1.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at sun.misc.Launcher$AppClassLoader.loadClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at com.adobe.air.apk.APKSigner.apkSign(APKSigner.java)
        at com.adobe.air.apk.APKOutputStream.d(APKOutputStream.java)
        at com.adobe.air.apk.APKPackager.renameOutput(APKPackager.java)
        at com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
        at com.adobe.air.ADT.parseArgsAndGo(ADT.java)
        at com.adobe.air.ADT.run(ADT.java)
        at com.adobe.air.ADT.main(ADT.java)
```

**Reply by SUSDEV** [reply]
So I next tried getting an early Java 8 JRE, I found 8_045 (closest to 41 you suggested). I compiled with that and it seemed to work both for arm7 and arm8. I have to test the resulting app, but I would call that big progress! Thanks so much.

**Reply by ajwfrost** [reply]
The apksigner tool is built with a requirement for Java 8 ... and we need to use apksigner in order to correctly sign the binary with the APK signature v2, required for Play Store uploads.

I guess it would be possible for us to check the current version of Java that's running and not do the APK signature if it's running 1.7, but generally the first 1.8 release is a good compromise between supporting the Java 8 language features (and Android tools) vs supporting the 1024-bit keys... 

So to help with the backwards compatibility a little, I think we should make two changes:
a) automatically use the 'legacy' build system, if the app descriptor namespace is less than 33.1. This will mean that people can continue to package their existing apps in the same way (but to use app bundle they'd need to update to 33.1, and once they're at 33.1 they have the option to switch to the legacy mode if they prefer)
b) check the Java runtime and don't try to load the APKSigner tool if we're before Java 8.. (I think that you'd still end up with a v1-signed APK file, from the jarsigner style mechanism)

thanks

**Reply by SUSDEV** [reply]
I don't know if it's time to give up Java 7 given that the Play Store requires APK signature v2. Are you thinking of trying to maintain Java 7 compatibility for apps not accessed through the Play Store? Having the app signed with v1 for me would be useless if the Play Store does not accept the update.

The early Java 8 build works with the 1024-bit certificates, so that's a small price to pay to get it working, along with updating the namespace to 33.1. I don't know how many developers are in my shoes, but it probably is fairly small. I think it's clear for me to keep doing updates beyond maybe this one I'm going to have to bite the bullet and try the aab route.

**Reply by ajwfrost** [reply]
If you're doing updates for the Play Store then you could go for the AAB route, the challenge there is that we use Android tools for this which now require Java 11. But then, the code-signing for an AAB is only the "jarsigner" which is like we had originally for APKs (jarsigner is the same as apk signing v1). Then it's a case of ensuring you have everything sorted out for the upload to the play store... per that tutorial doc, there are a few options here. If you've got an example of how to get it to work with an old certificate, it would probably be very useful to others :-) as I know a reasonable number who have still been contacting us about 1024-bit certificates and having problems with Java versions...

thanks
